scripts/generate/public_git.sh

```diff
@@ -0,0 +1,92 @@
+#!/bin/sh
+#
+# Generates support file from a public git repo
+
+set -e
+
+branch="master"
+directive_map=""
+match_func=""
+url=""
+path=""
+
+help() {
+    cat << EOF
+Generate support file from a git repo.
+
+usage: $(basename "$0") [-d|--directive-map] [-b|--branch] [--url] [-m|--match-func] [-o|--override] [-p|--path] [h|--help]
+    -h  | --help               Display this message
+    -d  | --directive-map      Name of the generated directive map(required)
+    -b  | --branch             Branch to checkout, defaults to "$branch"(optional)
+    -m  | --match-func         Name of the generated MatchFunc(required)
+    -p  | --path               The path of the output file(required)
+    --url                      Url used for git clone(required)
+EOF
+}
+
+while [ ! $# -eq 0 ]; do
+    case "$1" in
+        --help | -h)
+            help
+            exit 0
+            ;;
+        --directive_map | -d)
+            directive_map="$2"
+            shift
+            ;;
+        --branch | -b)
+            branch="$2"
+            shift
+            ;;
+        --match_func | -m)
+            match_func="$2"
+            shift
+            ;;
+        --url)
+            url="$2"
+            shift
+            ;;
+        --path | -p)
+            path="$2"
+            shift
+            ;;
+        *)
+            printf "Invalid option: %s\n" "$1" 1>&2
+            exit 1
+            ;;
+    esac
+    shift
+done
+
+if [ "$url" = "" ]; then
+    echo "url can't be empty"
+    exit 1
+fi
+
+if [ "$path" = "" ]; then
+    echo "path can't be empty"
+    exit 1
+fi
+
+if [ "$directive_map" = "" ]; then
+    echo "directive-map can't be empty"
+    exit 1
+fi
+
+if [ "$match_func" = "" ]; then
+    echo "match-func can't be empty"
+    exit 1
+fi
+
+echo "start generate from $url"
+tmp=$(mktemp -d)
+git clone "$url" "$tmp" --depth 1 --branch "$branch"
+
+generate_command="go run ./cmd/generate --src-path=$tmp --directive-map=$directive_map --match-func=$match_func"
+generate_command="$generate_command > $path"
+
+eval "$generate_command"
```

Comment:
Generally should avoid eval unless there is a good reason to.

```suggestion
go run ./cmd/generate "--src-path=$tmp" "--directive-map=$directive_map" "--match-func=$match_func" > $path
```
should do just fine, no?

If you want to make this a little neater and avoid long lines the pattern I like (requires /bin/bash at the top) is:
```suggestion
genArgs=(
  "--src-path=$tmp"
  # line comments are ok here to describe what you're doing
  "--directive-map=$directive_map"
  "--match-func=$match_func"
)

go run ./cmd/generate "${genArgs[@]}" > "$path"
```

genArgs is a bash array that we then splat (with quoting intact) onto the run command. The really nice thing about this array construct is that it doesn't require the `\` to continue lines that makes comments not work and is error prone as you add/remove lines.

Reply:
Another benefit of the bash array is that you can append to it, e.g. `genArgs+=(--directive-map=foo)`

With that, and the fact that it looks like a lot of the args are being passed through to cmd/generate can you do more of the error handling in there?  

E.g. if the the case statement was
```bash
    --match_func | -m)
            genArgs+=("--match-func=$2")
            shift
            ;;
```
and if the cmd/generate doesn't get a match function _it_ raises the error?

Reply:
> genArgs is a bash array that we then splat (with quoting intact) onto the run command.

Cool! I didn't know that before. Thanks for telling me this. You are right, the reason I used `eval` was I don't want long line and bash array seems a perfect solution. I will applied it.

>With that, and the fact that it looks like a lot of the args are being passed through to cmd/generate can you do more of the error handling in there?
E.g. if the the case statement was
    --match_func | -m)
            genArgs+=("--match-func=$2")
            shift
            ;;
and if the cmd/generate doesn't get a match function it raises the error?

I thought I already did the check here? Just after the arguments reading part

```bash
if [ "$url" = "" ]; then
    echo "url can't be empty"
    exit 1
fi

if [ "$path" = "" ]; then
    echo "path can't be empty"
    exit 1
fi

if [ "$directive_map" = "" ]; then
    echo "directive-map can't be empty"
    exit 1
fi

if [ "$match_func" = "" ]; then
    echo "match-func can't be empty"
    exit 1
fi
```

Reply:
yeah the array-splat is one of my favorite bash tricks, e.g. look at the use of "opts" separately in most functions in https://gitlab.com/f5/nginx/nginxazurelb/azure-infrastructure/-/snippets/2371626